sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -358,6 +360,35 @@ final class DataFrameWriter private[sql](df: DataFrame) {
   }
 
   /**
+   * :: Experimental ::
+   * Starts the execution of the streaming query, which will continually send results to the given
+   * [[ForeachWriter]] as new data arrives. The returned [[ContinuousQuery]] object can be used to
+   * interact with the stream.
+   *
+   * @since 2.0.0
+   */
+  @Experimental
+  def foreach(writer: ForeachWriter[T]): ContinuousQuery = {
+    val queryName = extraOptions.getOrElse("queryName", StreamExecution.nextName)
+    val checkpointLocation = extraOptions.get("checkpointLocation")
+      .orElse {
+        df.sparkSession.sessionState.conf.checkpointLocation.map { l =>
+          new Path(l, queryName).toUri.toString
+        }
+      }.getOrElse {
+      throw new AnalysisException("checkpointLocation must be specified either " +
+        "through option() or SQLConf")
+    }
+
+    df.sparkSession.sessionState.continuousQueryManager.startQuery(
```

Comment:
We have to clean the closure of the Writer and check for its serializability. Otherwise it might capture unexpected stuff in the closure and fail later.